app/org/sagebionetworks/bridge/dynamodb/DynamoUploadSchemaDao.java

```diff
@@ -136,125 +156,297 @@ public void setStudyIdIndex(DynamoIndexHelper studyIdIndex) {
+        List<UploadFieldDefinition> newFieldDefList = new ArrayList<>();
         for (SurveyQuestion oneQuestion : survey.getUnmodifiableQuestionList()) {
-            String name = oneQuestion.getIdentifier();
-            UploadFieldType type = getFieldTypeFromConstraints(oneQuestion.getConstraints());
-
-            // All survey questions are skippable, so mark the field as optional (not required)
-            UploadFieldDefinition oneFieldDef = new DynamoUploadFieldDefinition.Builder().withName(name)
-                    .withType(type).withRequired(false).build();
-            fieldDefList.add(oneFieldDef);
+            addFieldDefsForSurveyQuestion(newFieldDefList, oneQuestion);
         }
 
-        // Get the current rev.
-        String studyId = studyIdentifier.getIdentifier();
+        // Create schema. No need to set rev or version unless updating an existing rev. Study is always taken care of
+        // by the APIs.
         String schemaId = survey.getIdentifier();
-        DynamoUploadSchema oldUploadSchema = getUploadSchemaNoThrow(studyId, schemaId);
-        int oldRev = 0;
-        if (oldUploadSchema != null) {
-            oldRev = oldUploadSchema.getRevision();
-
-            // Optimization: If the new schema and the old schema have the same fields, return the old schema instead
-            // of creating a new one.
-            //
-            // Dump the fieldDefLists into a set, because if we have the same fields in a different order, the schemas
-            // are compatible, and we should use the old schema too.
-            Set<UploadFieldDefinition> oldFieldDefSet = ImmutableSet.copyOf(oldUploadSchema.getFieldDefinitions());
-            Set<UploadFieldDefinition> newFieldDefSet = ImmutableSet.copyOf(fieldDefList);
-            if (oldFieldDefSet.equals(newFieldDefSet)) {
-                return oldUploadSchema;
+        DynamoUploadSchema schemaToCreate = new DynamoUploadSchema();
+        schemaToCreate.setFieldDefinitions(newFieldDefList);
+        schemaToCreate.setName(survey.getName());
+        schemaToCreate.setSchemaId(schemaId);
+        schemaToCreate.setSchemaType(UploadSchemaType.IOS_SURVEY);
+        schemaToCreate.setSurveyGuid(survey.getGuid());
+        schemaToCreate.setSurveyCreatedOn(survey.getCreatedOn());
+
+        // Validate schema. There's nowhere else in the call path that validates it, so we have to do it here.
+        Validate.entityThrowingException(UploadSchemaValidator.INSTANCE, schemaToCreate);
+
+        // Short-cut: If we specify we want a new schema rev, then skip directly to creating a new schema rev.
+        if (!newSchemaRev) {
+            // Get the current rev.
+            String studyId = studyIdentifier.getIdentifier();
+            DynamoUploadSchema oldUploadSchema = getUploadSchemaNoThrow(studyId, schemaId);
+            if (oldUploadSchema != null && oldUploadSchema.getSchemaType() == UploadSchemaType.IOS_SURVEY) {
```

Comment:
Are you attempting to update the existing schema only if it's a survey schema? Do you plan to try this for task schemas as well, as client devs have tried to make fields optional before for these as well.